# Nobori_commensal_co-transcriptome_2020

Data availability
github
  normalized RNA-seq data for each strain with DEG info (/Users/tatsuyanobori/Desktop/MPIPZ/Projects/Bacterial_Transcriptome/A346_commensal_rnaseq_for_paper/rnaseq_summary_each_strain)
  PA list (Levy et al). Here or as Data S5.
  KO database for each strain + Ruben's database(maybe as Data S*)
  data required for bac-plant correlation analysis ("A310_tmm_OG-KO_FC_strain_combined.txt", "A347_fitted_mean_ave1_DEG_atleast_one_strain.txt")
  commensal genome files and gff files (gff gives cds*** count). make [cds***-OG-KO] table for each strain (/Users/tatsuyanobori/Desktop/MPIPZ/Projects/Bacterial_Transcriptome/A310_psl_commensal_rnaseq_analysis/A310_OG_analysis_3/A310_OG_annotated/annotation_combine_file_for_sharing). They are sufficient to make all the basic datasets?


Code availability
github
  ko_plot function
  ko_enrichment function
  config file (clustering function may be removed)
  KO_database_for_each_strain function
  bac-plant correlation bootstrap analysis
